feature/amazon/src/main/java/com/revenuecat/purchases/amazon/AmazonBilling.kt

```diff
@@ -140,31 +187,187 @@ class AmazonBilling constructor(
+            onError = {
+                completion(AmazonQueryPurchasesResult(isSuccessful = false, purchasesByHashedToken = emptyMap()))
+            }
+        )
+    }
+
+    // AmazonBilling delegates functionality to interfaces that have a common parent interface, it will only
+    // compile as long as all of the functions are implemented, otherwise it doesn't know which delegated
+    // implementation to take
+    override fun onUserDataResponse(response: UserDataResponse) {
+        userDataHandler.onUserDataResponse(response)
+    }
+
+    override fun onProductDataResponse(response: ProductDataResponse) {
+        productDataHandler.onProductDataResponse(response)
+    }
+
+    override fun onPurchaseResponse(response: PurchaseResponse) {
+        purchaseHandler.onPurchaseResponse(response)
+    }
+
+    override fun onPurchaseUpdatesResponse(response: PurchaseUpdatesResponse) {
+        purchaseUpdatesHandler.onPurchaseUpdatesResponse(response)
+    }
+
+    private fun logErrorsIfAny(errors: Map<String, PurchasesError>) {
+        if (errors.isNotEmpty()) {
+            val receiptsWithErrors = errors.keys.joinToString("\n")
+            log(
+                LogIntent.AMAZON_ERROR,
+                AmazonStrings.ERROR_FETCHING_RECEIPTS.format(receiptsWithErrors)
+            )
+        }
+    }
+
+    private fun getMissingSkusForReceipts(
+        appUserID: String,
+        amazonUserID: String,
+        receipts: List<Receipt>,
+        onCompletion: (tokensToSkusMap: Map<String, String>, errors: Map<String, PurchasesError>) -> Unit
+    ) {
+        val currentlyCachedTokensToSkus: Map<String, String> = cache.getReceiptSkus()
+
+        val successMap: MutableMap<String, String> = currentlyCachedTokensToSkus.toMutableMap()
+        val errorMap: MutableMap<String, PurchasesError> = mutableMapOf()
+
+        val nonSubscriptionReceiptsToSku =
+            receipts.filterNot { it.productType == ProductType.SUBSCRIPTION }
+                .map { it.receiptId to it.sku }
+
+        successMap.putAll(nonSubscriptionReceiptsToSku)
+
+        val subscriptionReceiptsToFetchTermSku: List<Receipt> =
+            receipts
+                .filter { it.productType == ProductType.SUBSCRIPTION }
+                .filterNot { currentlyCachedTokensToSkus.containsKey(it.receiptId) }
+
+        if (subscriptionReceiptsToFetchTermSku.isEmpty()) {
+            onCompletion(successMap, errorMap)
+            return
+        }
+
+        var receiptsLeft = subscriptionReceiptsToFetchTermSku.count()
+        subscriptionReceiptsToFetchTermSku.forEach { receipt ->
+            amazonBackend.getAmazonReceiptData(
+                receipt.receiptId,
+                amazonUserID,
+                onSuccess = { response ->
+                    log(LogIntent.DEBUG, AmazonStrings.RECEIPT_DATA_RECEIVED.format(response.toString()))
+
+                    successMap[receipt.receiptId] = response["termSku"] as String
+
+                    receiptsLeft--
```

Comment:
how does `receiptsLeft` work here in terms of getting referenced in the block? I'm asking because I believe in Swift / Objc it would actually get copied into each block, which would be a problem.

Reply:
In Java, `receiptsLeft` would need to be final. In Kotlin, it gets wrapped in a `IntRef`, which is final, but its value can be modified. This answer has a good explanation https://stackoverflow.com/a/44540217 

Do you like Kotlin even more now?

Reply:
I'm a believer 🙌